### PR TITLE
AB#92 Backend Persistence Handling

### PR DESCRIPTION
**Pull request by michaelschufi**
This is a prototype of the backend implementation.  The idea is that the client will send its past connectionId when initiating a connection. The server then looks it up or generates a new one.

I'm thinking of merging the connectionId and userId together, so everything is a bit clearer. What do you think?

### CODE REVIEW

**Comment by michaelschufi** [comment]
I will wait for feedback until I create and fix the tests.

File: game-backend/src/MainController.ts [hunk 1]
```diff
@@ -41,6 +44,32 @@ export default class MainController {
...
+  }
+
+  private static sendConnectionId(ws: WebSocket, connectionId: string): void {
+    ws.send(
+      JSON.stringify({
+        type: 'connectionInitialization',
+        payload: {
+          connectionId,
+        },
+      } as ConnectionInitialization)
```
Comment (StuckiSimon) [inline game-backend/src/MainController.ts]: Try not to do type casting - doing it using `as` will erase the benefit of using TypeScript.

For a reference implementation see:
https://github.com/code-and-conquer/darwin/blob/ae21189a07d11a4592aff2776b90d0ac64a03d72/game-frontend/src/components/user-script/index.tsx#L19

File: game-backend/src/MainController.ts [hunk 2]
```diff
@@ -41,6 +44,32 @@ export default class MainController {
...
 
+  /**
+   * Looks up the connection id in the store.
+   * It returns the found connectionId or generates a new one respectively.
+   * @param requestedConnectionId The connection id the client requests
+   */
+  private getConnectionId(requestedConnectionId: string): ConnectionId {
+    if (
+      requestedConnectionId &&
+      this.store.userContexts.userContextIds.indexOf(requestedConnectionId)
```
Comment (StuckiSimon) [inline game-backend/src/MainController.ts]: This probably won't work as expected in case the `requestedConnectionId` is at `userContextIds[0]`

**Review by StuckiSimon — COMMENTED** [review]
Just some general feedback for your wip.
Unit Tests don't like it 😢

File: game-backend/src/MainController.ts (hunk 2 above)
Comment (michaelschufi) [reply game-backend/src/MainController.ts]: Thanks!

File: game-backend/src/MainController.ts (hunk 1 above)
Comment (michaelschufi) [reply game-backend/src/MainController.ts]: Thank you, somehow I always mess that up...

File: darwin-types/messages/ConnectionInitialization.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import { Message } from './Message';
+import { ConnectionId } from '../../game-backend/src/ServerStore';
```
Comment (StuckiSimon) [inline darwin-types/messages/ConnectionInitialization.ts]: Could you extract the ConnectionId to darwin-types? 😇

**Review by StuckiSimon — CHANGES_REQUESTED** [review]
small import comment

**Review by StuckiSimon — APPROVED** [review]
Looks fine for me.
The only drawback is the fact that anyone can impersonate every user on the system due to the predictable nature of hyperIds.

**Comment by michaelschufi** [comment]
That's true but is not a requirement for now.

**Pull request merged**